src/rust/ide/lib/view/src/graph_editor/component/edge.rs

```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
+/// otherwise coloring it later will also color the hover area.
+fn hover_area(base_shape:AnyShape, size:Var<Pixels>) -> AnyShape {
+    let hover_area = base_shape.grow(size).fill(color::Rgba::new(0.0,0.0,0.0,0.000_001));
+    (hover_area + base_shape).into()
+}
+
+
+
+// ==================
+// === FocusSplit ===
+// ==================
+
+/// Holds the data required to split a shape into two focus visual groups.
+#[derive(Clone,Copy,Debug)]
+struct FocusSplit {
+    position : Vector2<f32>,
+    angle    : f32
+}
+
+impl FocusSplit {
+    fn new(position:Vector2<f32>, angle:f32) -> Self {
+        FocusSplit {position,angle}
+    }
+}
+
+
+
+// ===================
+// === FocusedEdge ===
+// ===================
+
+/// An edge split into two parts - focused and unfocused one.
+struct FocusedEdge {
+    focused   : AnyShape,
+    unfocused : AnyShape,
+}
+
+impl FocusedEdge {
+    /// Splits the shape in two at the line given by the `split_center` and `split_angle`.
+    fn new
+    ( base_shape   : impl Into<AnyShape>
+    , split_center : &Var<Vector2<Pixels>>
+    , split_angle  : &Var<Radians>
+    ) -> Self {
+        let base_shape = base_shape.into();
+        let split_mask = HalfPlane().rotate(split_angle).translate(split_center);
+        let focused    = (&base_shape * &split_mask).into();
+        let unfocused  = (&base_shape - &split_mask).into();
+        FocusedEdge {focused,unfocused}
+    }
+
+    /// Color the focused and unfocused parts with the provided colors.
+    fn fill<C:Into<color::Rgba>>(&self, focused_color:C, unfocused_color:C) -> AnyShape {
+        let focused_color   = focused_color.into();
+        let unfocused_color = unfocused_color.into();
+        let focused         = self.focused.fill(&focused_color);
+        let unfocused       = self.unfocused.fill(&unfocused_color);
+        (focused + unfocused).into()
+    }
+}
+
+
+
+// ==================
+// === SnapTarget ===
+// ==================
+
+/// `SnapTarget` is the result value of snapping operations on `AnyEdgeShape`. It holds the
+/// shape that a hover position was snapped to and the snapped position on the shape. The snapped
+/// position lies (a) on the visible part of the shape and (b) is the closes position on the shape
+/// to the source position that was used to compute the snapped position.
+#[derive(Clone,Debug)]
+struct SnapTarget {
+    position        : Vector2<f32>,
+    target_shape_id : display::object::Id,
+}
+
+impl SnapTarget {
+    fn new(position:Vector2<f32>, target_shape_id:display::object::Id) -> Self {
+        SnapTarget {position,target_shape_id}
+    }
+}
+
 
 
 // =========================
 // === Shape Definitions ===
 // =========================
 
-macro_rules! define_corner_start {($($color:tt)*) => {
+/// Joint definition
+pub mod joint {
+    use super::*;
+
+    ensogl::define_shape_system! {
+            (
+            ) {
+                let radius : Var<Pixels> = "min(input_size.y,input_size.x)".into();
```

Comment:
wouldn't these be always the same ? Is there any place where we keep the input_size not rectangular? If no, let's not waste gpu computing power ;)